symex: source is not optional

kroening opened the pull request:
This is probably the worst case of use of unitialised data that we have. Most parts of the code base assume that source.pc is a valid iterator, yet the interface explicitly allows it to be uninitialised.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

kroening committed:
remove the trivial constructor from symex_targett::sourcet

sourcet contains an interator, and the current interface explicitly allows
it to be uninitialised.  However, most parts of the code base access
source.pc without checking.  This commit removes the option to leave the
source uninitialized.

kroening committed:
Ensure goto_symex_statet::source is always set

sourcet contains an interator, and the interface explicitly allows it to be
uninitialised.  However, most parts of the code base access source.pc
without checking.

tautschnig reviewed (approved):
I approve of the general direction, but I would suggest that the two commits be squashed (and some further work is necessary as CI is failing) as neither of them makes sense without the other.

kroening commented:
The first one works by itself, and could be seen as an improvement.
The second then removes dead code.
The CI issue is that slice_by_trace wants to produce SSA_steps without source, and there isn't an obvious one to use.

tautschnig commented:
> The CI issue is that slice_by_trace wants to produce SSA_steps without source, and there isn't an obvious one to use.

One could be solved by storing a source in the `merge_map_back`, the other could probably just use the source from the front element (before emplacing a new element at the front)?

tautschnig commented:
This one should be easy once #3971 has been merged.

kroening commented:
> This one should be easy once #3971 has been merged.

Indeed!

tautschnig commented:
Rebased as #3971 has been merged.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 2cd585a2ffd3f810a94c49099da7800234bee370).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99243672

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 9d3d87bd631f7f9ca30b77d8e1b4bb01443e3e79).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99851584

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 58d7bcf1dad74eaac2e2996e8895674fdc2b8d2f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100326162

tautschnig committed:
Refactor goto_symext initialisation API

Avoid doing the same start_function set-up three times, and avoid redundant work
between different initialize* functions when one always calls the other.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 9e0c5a3d5170eb1b2160a9685023bee5816b6644).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100660772
